In [12]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Conv2D
from tensorflow.keras.layers import MaxPooling2D

## Task 1 — Build and Inspect a Convolutional Network
No training required; the goal is solely to construct the network.
- Import `Sequential`, `Input`, and `Conv2D`.
- Build a model that accepts a `(28, 28, 1)` image and contains a single `Conv2D(1, (3, 3))` layer.
- Run `model.summary()`. Read the parameter count and output shape.
- Change the layer to `Conv2D(8, (3, 3))`, then change the input shape to `(28, 28, 3)`. Record the parameter count each time.
- Answer: Where does the parameter count come from in each of the three cases? Why did the 28×28 image shrink to 26×26? Why does the parameter count depend on the number of channels rather than the image dimensions (28×28)?

In [3]:
Model_1 = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(filters=1, kernel_size=(3, 3))
    ])
Model_1.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_1 (Conv2D)               │ (None, 26, 26, 1)      │            10 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 10 (40.00 B)

 Trainable params: 10 (40.00 B)

 Non-trainable params: 0 (0.00 B)

 (3 * 3 * 1 + 1) * 1 = 10 - params

In [4]:
Model_2 = Sequential([
    Input(shape=(28, 28, 3)),
    Conv2D(filters=8, kernel_size=(3, 3))
    ])
Model_2.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_2 (Conv2D)               │ (None, 26, 26, 8)      │           224 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 224 (896.00 B)

 Trainable params: 224 (896.00 B)

 Non-trainable params: 0 (0.00 B)

(3 * 3 * 3 + 1) * 8 = 224 - params

# Answers

1. **Parameter calculation formula:**
   `params = (kernel_h * kernel_w * input_channels + 1_bias) * num_filters`
   - **Case 1:** `(3 * 3 * 1 + 1) * 1 = 10`
   - **Case 2:** `(3 * 3 * 3 + 1) * 8 = 224`
2. **Why 28×28 shrinks to 26×26:**
   Default padding is `padding="valid"` (no zero-padding). 
   Output size formula: `(W - K + 2P(padding =0) / S + 1 = (28 - 3 + 0) / 1(stride usually -1) + 1 = 26`.
   A 3×3 kernel cannot center on the outermost border pixels without extending beyond image boundaries, losing 1 pixel oneach edge (2 total).
3. **Why parameter count depends on channels, not spatial resolution (28×28):**
   **Weight sharing.** A convolutional filter slides across the entire spatial area using the exact same set of weights regardless of image height or width. However, each filter must span the entire depth of the input, meaning weights must be defined for every input channel.

# Task 2 — Padding, Stride, and Pooling


Still no training required. The objective is to examine how different layers alter the shape of the data.

Using an input shape of `(28, 28, 1)`, sequentially build models with the following layer configurations:
1. `Conv2D(8, (3, 3), padding='valid')`
2. `Conv2D(8, (3, 3), padding='same')`
3. `Conv2D(8, (3, 3), strides=2, padding='same')`
4. `Conv2D(8, (5, 5))`
5. `Conv2D(8, (3, 3))` followed by `MaxPooling2D((2, 2))`

For each configuration, record the **output shape** and the **parameter count**.

### Questions to answer:
- What is the difference between `valid` and `same` padding?
- Why does `MaxPooling2D` have 0 parameters?
- How does downsampling via `strides=2` differ from downsampling via pooling?

In [7]:
Model_3 = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(filters=8, kernel_size=(3, 3), padding='valid')
    ])
Model_3.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)               │ (None, 26, 26, 8)      │            80 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 80 (320.00 B)

 Trainable params: 80 (320.00 B)

 Non-trainable params: 0 (0.00 B)

(3 * 3 * 1 + 1) * 8 = 80

In [8]:
Model_4 = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(filters=8, kernel_size=(3, 3), padding='same')
    ])
Model_4.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_4 (Conv2D)               │ (None, 28, 28, 8)      │            80 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 80 (320.00 B)

 Trainable params: 80 (320.00 B)

 Non-trainable params: 0 (0.00 B)

- **`padding='same'`**: Zero-pads borders to preserve spatial dimensions (does not affect parameter count).
- **`strides=2`**: Steps by 2 pixels to downsample dimensions by half using learnable convolutional weights.
- **`MaxPooling2D`**: Extracts the maximum value in each window to downsample with no learnable weights (always 0 params).

In [9]:
Model_5 = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(filters=8, kernel_size=(3, 3), padding='same', strides=2)
    ])
Model_5.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_5 (Conv2D)               │ (None, 14, 14, 8)      │            80 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 80 (320.00 B)

 Trainable params: 80 (320.00 B)

 Non-trainable params: 0 (0.00 B)

shape 28 /2 = 14

In [10]:
Model_6 = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(filters=8, kernel_size=(5, 5))
    ])
Model_6.summary()

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_6 (Conv2D)               │ (None, 24, 24, 8)      │           208 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 208 (832.00 B)

 Trainable params: 208 (832.00 B)

 Non-trainable params: 0 (0.00 B)

(5 * 5 * 1 +1) * 8 = 208

In [13]:
Model_7 = Sequential([
    Input(shape=(28, 28, 1)),
    Conv2D(filters=8, kernel_size=(3, 3)),
    MaxPooling2D(pool_size=(2, 2))
])
Model_7.summary()

Model: "sequential_7"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_7 (Conv2D)               │ (None, 26, 26, 8)      │            80 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 8)      │             0 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 80 (320.00 B)

 Trainable params: 80 (320.00 B)

 Non-trainable params: 0 (0.00 B)

1. What is the difference between valid and same padding?

    valid: No zero-padding applied. The kernel only moves strictly within image boundaries, causing spatial dimensions to shrink ($28\times 28 \to 26\times 26$).
    
    same: Pads the image perimeter with zeros. This preserves identical input spatial dimensions at the output ($28\times 28 \to 28\times 28$ when stride is 1).


2. Why does MaxPooling2D have 0 parameters?
    
    MaxPooling is a fixed, deterministic mathematical operation ($\max$), not a learnable layer. It simply extracts the highest numerical value in each window ($2\times 2$)—there are no weights ($w$) or biases ($b$) to update via backpropagation.

3. How does downsampling via strides=2 differ from downsampling via pooling?

    strides=2 (convolution): Downsamples through learnable weights. The model learns how best to summarize and compress spatial information while extracting features.
    
    MaxPooling: Downsamples through a fixed heuristic. It does not learn; it simply retains the strongest activation in each region while discarding the rest.

# Task 3 — Teach the Network a Single Filter


The 2D equivalent of $y = 2x + 1$. The model's objective is to discover a known ground-truth filter.

- Define a vertical edge-detection filter: `kernel = [[1, 0, -1], [1, 0, -1], [1, 0, -1]]`.
- Generate ~500 random 10×10 images (`np.random.rand`). For each image, calculate $y$ using `scipy.signal.correlate2d(image, kernel, mode='valid')`.
- Build a model with a single `Conv2D(1, (3, 3))` layer, compile it (`optimizer='adam'`, `loss='mse'`), and train it.
- Extract `model.get_weights()`. Reshape the learned weights to `(3, 3)` and compare them with the target `kernel`. What is the value of the bias?

### Questions to answer:
- Why was the network able to reconstruct the filter almost perfectly?
- What does this demonstrate about what a convolutional layer actually learns?